File: 1.GettingStarted/questions.java
import java.lang.Math;
import java.util.Scanner;

public class questions {

    public static boolean is_prime(int n)
    {
        for (int i = 2; i * 2 <= n;i++)
        {
            if (n%i == 0)
            {
                return false;
            }
        }
        return true;
    }

    public static void print_all_primes(int n)
    {
        int numbers[] = new int[n+1];
        for (int idx = 0; idx <= n ;idx++)
        {
            numbers[idx] = 1;
        }
        for(int current_idx=2; current_idx<=n;current_idx++)
        {
            if (numbers[current_idx] == 1)
            System.out.println(current_idx);
            for(int idx = 2; ;idx++)
            {
                if (current_idx*idx > n)
                break;
                numbers[current_idx*idx] = 0;
            }
        }

    }

    public static void reverse_of_a_number(int n)
    {
        int val = n;
        while (val > 0)
        {
            int divisor = val%10;
            val = val/10;
            System.out.println(divisor);
        }
    }

    public static void digits_of_a_number(int n)
    {
        int len = 0;
        int val = n;
        while(val >0)
        {
            len++;
            val = val /10;
        }
        int remainder = n;
        int divisor = (int) Math.pow(10,len - 1);
        while(len > 0)
        {
            System.out.println(remainder/divisor);
            remainder = remainder % divisor;
            len--;
            divisor = divisor/10;
        }
    }

    public static int inverse_of_a_number(int n)
    {
        int number = n;
        int multiplier = 10;
        int inverted_number = 0;
        int idx = 1;

        while(number > 0)
        {
            // traverse number from end i.e index 1
            // multiplier = power(10,1 - 1)
            // inverted_number = inverted_number + (digit*multiplier)
            // number = number / 10
            int digit = number % 10;
            multiplier = (int) Math.pow(10,digit - 1);
            inverted_number = inverted_number + (idx*multiplier);
            number = number / 10;
            idx++;

        }

        return inverted_number;
    }
    
    public static int get_num_length(int num)
    {
        int number = num;
        int len = 0;
        while(number>0)
        {
            len++;
            number = number/10;
        }
        return len;
    }

    public static int rotate_a_number(int num,int k)
    {   
        int number = num;
        int number_len = get_num_length(number);
        if(k<0)
        {
            k = number_len + k;
        }
        int multiplier = (int) Math.pow(10,number_len - k);
        int divider = (int) Math.pow(10,k);
        // 12345 -> 2 -> 45123
        // 12345 -> -2 -> 34512
        // 1. get multiplier and divider
        // 2. get right_digits*multiplier + left_digits/divider 
        int left_dig = number/divider;
        int right_dig = (number % divider)*multiplier;
        int rotated_num = right_dig + left_dig;
        return rotated_num;
    }

    public static int greatestCommonDiviser(int n1,int n2)
    {
        int gcd = 1;

        // 14 , 32, 34,42
        int divident = n1 ;
        int divisor = n2;
        int remainder = divident % divisor;
        while(remainder>0)
        {
            divident = divisor;
            divisor = remainder;
            remainder = divident % divisor;
        }
        gcd = divisor;
        return gcd;
    }


    public static void temp(int n)
    {
        n = n + 1;
    }


    public static void main(String args[])
    {
        // int n = 11;
        // System.out.println(is_prime(n));

        // print_all_primes(30);

        // digits_of_a_number(123456);
        // reverse_of_a_number(123456);
        // System.out.println(inverse_of_a_number(34152));
        // int n = 5;
        // temp(n);
        // Scanner scn = new Scanner(System.in);
        // int k = scn.nextInt();
        // System.out.println(rotate_a_number(12345,-2));
        // System.out.println(greatestCommonDiviser(14,37));
        // primeFactorization(14);
        
    }
}


// 7. Is a Prime Number
// 8. Print All Primes till Number

// 9. Digits of a Number
// 10. Reverse of a Number
// 11. Inverse a Number
// 12. Rotate a Number

// 13. GCD and LCM 
// 14. Prime Factorization of a Number

// 15. Pythagorean Factorization of a Number
// 16. The Curious Case of Benjamin Bulbs

File: questions.py



def is_prime(num):
    i = 2
    while(i*i < num):
        i += 1
        if (num % i == 0):
            print("false")
            return
    print("true")

def print_all_primes_till_number(num):
    # get a list of all 1s
    # traverse the list from 2
    # if list[idx] is 1 print and mark its multiples as 0 in list
    list1 = [1]*(num + 1)
    list1[0] = list1[1] = 0
    for idx in range(2,num + 1):
        idx2 = idx
        if (list1[idx] == 1):
            print(idx)
            while (idx2 <= num):
                list1[idx2] = 0
                idx2 += idx
            
def get_num_len(number):
    len = 0
    while(number > 0):
        len += 1
        number = number//10
    return len


def digits_of_number(number):
    num_length = get_num_len(number)
    divisor = pow(10,num_length - 1)
    while(divisor > 0):
        digit = number//divisor
        number = number - (digit*divisor)
        print(digit)
        divisor = divisor//10

def reverse_of_number(number):
    while(number > 0):
        digit = number % 10
        print(digit)
        number //= 10

def inverse_of_number(number):
    inversed_number = 0
    idx = 1
    while(number > 0):
        digit = number % 10
        multiplier = pow(10,digit - 1)
        inversed_number = inversed_number + (idx*multiplier)
        number //= 10
        idx += 1
    print(inversed_number)


def rotate_number(number,k):
    number_length = get_num_len(number)
    k = (number_length + k)  % number_length
    # 5 -2  = 3 % 5 = 3
    # 5 + 2 = 7 % 5 = 2
    #  split number -> left + (right*multiplier)
    divisor = pow(10,k)
    multiplier = pow(10,number_length - k)
    right = number % divisor
    left = number // divisor
    rotated_number = left + (right*multiplier)
    print(rotated_number)

def gcd(number1,number2):
    divisor = number1
    divident = number2
    remainder = divisor
    while (remainder > 0):
        divisor = remainder
        remainder = divident % divisor
        divident = divisor
        print(divident , "/" , divisor , "=" , remainder)
    return divisor

def lcm(num1,num2):
    num_gcd = gcd(num1,num2)
    print(num1*num2/num_gcd)

def printj():
    j = 2
    interval = 2
    for j in range(1,15,2):
        print("idx : " ,j)

num = 72
# is_prime(num)
# print_all_primes_till_number(num)
# printj()
# digits_of_number(2345)
# reverse_of_number(2356)
# inverse_of_number(15243)
# rotate_number(12345,2)
# print(gcd(12,17))
# lcm(12,24)



# 7. Is a Prime Number
# 8. Print All Primes till Number
# 9. Digits of a Number
# 10. Reverse of a Number
# 11. Inverse a Number
# 12. Rotate a Number
# 13. GCD and LCM 
# 14. Prime Factorization of a Number
# 15. Pythagorean Factorization of a Number
# 16. The Curious Case of Benjamin Bulbs